src/libsyntax/attr/mod.rs

```diff
@@ -275,17 +301,42 @@ impl AttrItem {
 }
 
 impl Attribute {
+    pub fn is_sugared_doc(&self) -> bool {
+        match self.kind {
+            AttrKind::Normal(_) => false,
+            AttrKind::SugaredDoc(_) => true,
+        }
+    }
+
+    pub fn get_normal_item(&self) -> &AttrItem {
+        match self.kind {
+            AttrKind::Normal(ref item) => item,
+            AttrKind::SugaredDoc(_) => panic!("unexpected sugared doc"),
+        }
+    }
+
+    pub fn unwrap_normal_item(self) -> AttrItem {
+        match self.kind {
+            AttrKind::Normal(item) => item,
+            AttrKind::SugaredDoc(_) => panic!("unexpected sugared doc"),
+        }
+    }
+
     /// Extracts the MetaItem from inside this Attribute.
     pub fn meta(&self) -> Option<MetaItem> {
-        self.item.meta(self.span)
+        match self.kind {
+            AttrKind::Normal(ref item) => item.meta(self.span),
+            AttrKind::SugaredDoc(comment) =>
+                Some(mk_name_value_item_str(Ident::new(sym::doc, DUMMY_SP), comment, DUMMY_SP)),
```

Comment:
```suggestion
                Some(mk_name_value_item_str(Ident::new(sym::doc, self.span), comment, self.span)),
```

Reply:
I originally had `self.span` in all the places you suggested, but changed it to `DUMMY_SP` because `self.span` is an approximation at best; there isn't really a correct span possible here. But if span approximation is generally used when desugaring things then I'm fine to continue that trend here.

Reply:
Approximation is certainly better than `DUMMY_SP`, all kinds of proc macros and derives span their output to the macro call unless there are better alternatives.
Also, the `SyntaxContext` component of the span is precise in this case.

Reply:
Thanks for the explanation. I will change this.